src/goto-programs/remove_const_function_pointers.h

```diff
@@ -0,0 +1,96 @@
+/*******************************************************************\
+
+Module: Goto Programs
+
+Author: Thomas Kiley, thomas.kiley@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_GOTO_PROGRAMS_REMOVE_CONST_FUNCTION_POINTERS_H
+#define CPROVER_GOTO_PROGRAMS_REMOVE_CONST_FUNCTION_POINTERS_H
+
+#include "goto_model.h"
+#include <util/message.h>
+#include <util/expr.h>
+#include <util/namespace.h>
+
+
+class remove_const_function_pointerst:public messaget
+{
+public:
+  typedef std::set<exprt> functionst;
+  typedef std::list<exprt> expressionst;
+  remove_const_function_pointerst(
+    message_handlert &message_handler,
+    const exprt &base_expression,
+    const namespacet &ns,
+    const symbol_tablet &symbol_table);
+
+  bool operator()(functionst &out_functions);
+
+private:
+  exprt replace_const_symbols(const exprt &expression);
+  exprt resolve_symbol(const symbol_exprt &symbol_expr);
+
+  // recursive functions for dealing with the function pointer
+  bool try_resolve_function_call(const exprt &expr, functionst &out_functions);
+
+  bool try_resolve_function_calls(
+    const expressionst &exprs, functionst &out_functions);
+
+  bool try_resolve_index_of_function_call(
+    const index_exprt &index_expr, functionst &out_functions);
+
+  bool try_resolve_member_function_call(
+    const member_exprt &member_expr, functionst &out_functions);
+
+  bool try_resolve_address_of_function_call(
+    const address_of_exprt &address_expr, functionst &out_functions);
+
+  bool try_resolve_dereference_function_call(
+    const dereference_exprt &deref_expr, functionst &out_functions);
+
+  bool try_resolve_typecast_function_call(
+    const typecast_exprt &typecast_expr, functionst &out_functions);
+
+  // recursive functions for dealing with the auxillary elements
+  bool try_resolve_expression(
+    const exprt &expr,
+    expressionst &out_resolved_expression,
+    bool &out_is_const);
+
+  bool try_resolve_index_of(
+    const index_exprt &index_expr,
+    expressionst &out_expressions,
+    bool &out_is_const);
+
+  bool try_resolve_member(
+    const member_exprt &member_expr,
+    expressionst &out_expressions,
+    bool &out_is_const);
+
+  bool try_resolve_dereference(
+    const dereference_exprt &deref_expr,
+    expressionst &out_expressions,
+    bool &out_is_const);
+
+  bool try_resolve_typecast(
+    const typecast_exprt &typecast_expr,
+    expressionst &out_expressions,
+    bool &out_is_const);
+
+  bool is_expression_const(const exprt &expression);
+  bool is_type_const(const typet &type);
```

Comment:
Likely those two functions should be marked const? Also, `is_const_expression`, `is_const_type` would seem more consistent (e.g., replace_const_symbol).